raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java

```diff
@@ -609,6 +708,161 @@ public void testBeginQuorumEpochTimer(boolean withDirectoryId) {
         assertEquals(0, state.timeUntilBeginQuorumEpochTimerExpires(time.milliseconds()));
     }
 
+    @Test
+    public void testVolatileVoters() {
+        int follower1 = 1;
+        long epochStartOffset = 10L;
+
+        VoterSet voters = localWithRemoteVoterSet(IntStream.of(follower1), false);
+        LeaderState<?> state = newLeaderState(
+            voters,
+            epochStartOffset,
+            KRaftVersion.KRAFT_VERSION_0
+        );
+
+        var votersWithLeaderUpdated = state.volatileVoters().get();
+        assertEquals(
+            voters.updateVoterIgnoringDirectoryId(localVoterNode).get(),
+            votersWithLeaderUpdated.voters()
+        );
+
+        var updatedVoters = new KRaftVersionUpgrade.Voters(
+            votersWithLeaderUpdated
+                .voters()
+                .updateVoterIgnoringDirectoryId(VoterSetTest.voterNode(follower1, true))
+                .get()
+        );
+
+        // Upate in-memory voter and check state
+        assertTrue(
+            state.compareAndSetVolatileVoters(votersWithLeaderUpdated, updatedVoters)
+        );
+        assertEquals(updatedVoters, state.volatileVoters().get());
+
+        // Unable to perform atomic update
+        assertFalse(
+            state.compareAndSetVolatileVoters(votersWithLeaderUpdated, updatedVoters)
+        );
+    }
+
+    @Test
+    public void testInvalidMaybeAppendUpgradedKRaftVersion() {
+        int follower1 = 1;
+        int follower2 = 2;
+        long epochStartOffset = 10L;
+
+        VoterSet persistedVoters = localWithRemoteVoterSet(IntStream.of(follower1, follower2), false);
+        LeaderState<?> state = newLeaderState(
+            persistedVoters,
+            epochStartOffset,
+            KRaftVersion.KRAFT_VERSION_0
+        );
+
+        assertThrows(
```

Comment:
nit, could you add short comments that explain why each is invalid?

Reply:
Fixed.